mkdir_linux.go

```diff
@@ -0,0 +1,224 @@
+		// If we hit the root, continue on without actually opening it.
+		if nextPath == "/" {
+			// At this point, currentDir must be the same as rootDir (there is
+			// only a single ".." component), so we can just update the logical
+			// path without changing the handle.
+			currentPath = nextPath
+			continue
+		}
+
+		// Try to open the next component as a directory.
+		nextDir, err := openatFile(currentDir, part, unix.O_PATH|unix.O_NOFOLLOW|unix.O_DIRECTORY|unix.O_CLOEXEC, 0)
+		switch {
+		case err == nil:
+			// If we are operating on a .., make sure we haven't escaped. We
+			// only have to check for ".." here because walking down into a
+			// regular component component cannot cause you to escape. This
+			// mirrors the logic in RESOLVE_IN_ROOT, except we have to check
+			// every ".." rather than only checking after a rename or mount on
+			// the system.
+			if part == ".." {
+				// Make sure the root hasn't moved.
+				if err := checkProcSelfFdPath(root, rootDir); err != nil {
+					return nil, "", fmt.Errorf("root path moved during lookup: %w", err)
+				}
+				// Make sure the path is what we expect.
+				fullPath := root + nextPath
+				if err := checkProcSelfFdPath(fullPath, nextDir); err != nil {
+					return nil, "", fmt.Errorf("walking into .. had unexpected result: %w", err)
+				}
+			}
+			_ = currentDir.Close()
+			currentDir = nextDir
+			currentPath = nextPath
+
+		case errors.Is(err, os.ErrNotExist):
+			// We have hit a final component that doesn't exist, so we have our
+			// partial open result. Note that we have to use the OLD remaining
+			// path, since the lookup failed.
+			return currentDir, oldRemainingPath, nil
+
+		case errors.Is(err, unix.ELOOP), errors.Is(err, unix.ENOTDIR):
+			// (O_PATH|O_NOFOLLOW) and O_DIRECTORY means we cannot tell if we
+			// hit a symlink or some other non-directory component from the
+			// error, but we have to readlink the target anyway so we can check
+			// that way so assume it's a symlink.
+			linkDest, err := readlinkatFile(currentDir, part)
+			if err != nil {
+				if errors.Is(err, unix.EINVAL) {
+					// The part was not a symlink, so assume that it's a
+					// regular file. It is possible for it to be a directory
+					// (if an attacker is swapping a directory and
+					// non-directory at this subpath) but erroring out here is
+					// better anyway.
+					err = fmt.Errorf("path component is invalid: %w", unix.ENOTDIR)
+				}
+				return nil, "", err
+			}
+
+			linksWalked++
+			if linksWalked > maxSymlinkLimit {
+				return nil, "", &os.PathError{Op: "partialOpen", Path: root + "/" + unsafePath, Err: unix.ELOOP}
+			}
+
+			remainingPath = linkDest + "/" + remainingPath
+			// Absolute symlinks reset any work we've already done.
+			if path.IsAbs(linkDest) {
+				nextDir, err := dupFile(rootDir)
+				if err != nil {
+					return nil, "", fmt.Errorf("clone root fd: %w", err)
+				}
+				_ = currentDir.Close()
+				currentDir = nextDir
```

Comment:
If this is an abs link, what guarantees that the returned fd will be inside the root? In particular, if `remainingPath = ""` and part is a link to `/`, for example.

Reply:
`nextDir` is a clone of `rootDir` here (I should give it a different name though, good point). Also, if you get here the current component was a symlink and so `remainingPath` cannot be `""`.